### ISSUE TITLE
Deprecate the validate method (used in forms)

### ISSUE THREAD

**Issue by eximius313** [playframework/playframework#5992]
### Play Version (2.5.x / etc)

2.5.1
### API (Scala / Java / Neither / Both)

Java
### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

N/A
### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

N/A
### Library Dependencies

N/A
### Expected Behavior
1. form validation method need's to use database repository or other DI service
2. Service is being injected into form using `@Inject`
### Actual Behavior
1. Using `@Inject` on a field is ignored
2. Using `@Inject` in constructor throws:`
   [CompletionException: java.lang.RuntimeException: Cannot instantiate class controllers.myController.forms.MyForm. It must have a default constructor]`
3. Using `Play.application().injector().instanceOf(MyService.class)` is deprecated

**Comment by gmethvin** [comment]
What does your form class look like?

**Comment by gmethvin** [comment]
It should technically be possible to inject forms but I think creating a custom validator and injecting into that is the preferred way.

**Comment by eximius313** [comment]
```
public class LoginForm {
	@Constraints.Email
	@Constraints.Required
	@Constraints.MaxLength(255)
	private String email;
	
	@Constraints.Required
	private String password;
	
	public List<ValidationError> validate() {
		List<ValidationError> result = new ArrayList<ValidationError>();
		email = email.toLowerCase().trim();
		UserRepository userRepository = Play.application().injector().instanceOf(UserRepository.class);
		boolean ok = userRepository.authenticate(email, password);
		if(ok == false){
			result.add(new ValidationError("email","bad.username.or.password"));
			result.add(new ValidationError("password","bad.username.or.password"));
		}
		return result.isEmpty() ? null : result;
	}
}
```

**Comment by marcospereira** [comment]
I don't think this is a good use of forms. They are very simple classes that transport data and can do simple validation (empty, well formed email, well formed url, etc) on this data. But I'm not sure if it is a good idea to have forms depending on external services like a database.

Your use case can be refactored into something like this:

```java
public class LoginForm {
    @Constraints.Email
    @Constraints.Required
    @Constraints.MaxLength(255)
    private String email;

    @Constraints.Required
    private String password;
}
```

Here your form is just a data transporter with simple validations (email, required and max length). And then your action would be:

```java
public Result authenticate() {
    Form<LoginForm> form = formFactory.form(LoginForm.class).bindFromRequest();
    if (form.hasErrors()) {
        return badRequest(views.html.login.render(form));
    }

    LoginForm loginForm = form.get();
    // now it is returning a user instead of a boolean
    User user = userRepository.authenticate(email, password);
    if (user == null) {
        flash("login.failed.message", "bad.username.or.password");
        return badRequest(views.html.login.render(form));
    }

    // add user info to the session
    // return successful result
}
```

Of course, both `userRepository` and `formFactory` were injected at the controller.

---------

@gmethvin if I remember correctly, there was some discussion at the gitter channel about the use of forms, but I couldn't find it. Not sure about what to do with this issue. Close it? Rewrite it to suggest some documentation change?

**Comment by gmethvin** [comment]
I misunderstood the issue. You mean it's not possible to inject members into the instance that backs your form. This is correct. The `validate` method only has access to the members of your class. It's possible we could change this, but I agree with @marcospereira that it's not really the correct way to use forms. 

If you did see a use case for using injected components in validation, you could probably define a class-level constraint and a custom validator for that that did what you want. I'm not totally familiar with the Java validation stuff so I couldn't tell you how to do it off the top of my head though.

**Comment by eximius313** [comment]
There are a lot of cases where gathering information from the database **is a part of validation** and form **should not** let you go further. Another use case is the use of MessagesApi (I want to return already localized message).
I could say that there are two approaches to this problem:
1)Leave the Form simple and do additional checks along with intended logic in controller
2)Leave validation where it should be - in Form and just do intended logic in controller
I believe that both approaches are valid and If you are forcing DI (which is good) please allow users to **use it where it's needed** and let them choose the approach.
If you leave us only option 1, you force us to rewrite dozens of forms and controllers against approach 2.

**Comment by eximius313** [comment]
There are a lot of cases where gathering information from the database **is a part of validation** and form **should not** let you go further. Another use case is the use of MessagesApi (I want to return already localized message).
I could say that there are two approaches to this problem:
1)Leave the Form simple and do additional checks along with intended logic in controller
2)Leave validation where it should be - in Form and just do intended logic in controller
I believe that both approaches are valid and If you are forcing DI (which is good) please allow users to **use it where it's needed** and let them choose the approach.
If you leave us only option 1, you force us to rewrite dozens of forms and controllers against approach 2.

**Comment by gmethvin** [comment]
@eximius313 Have you tried using a class-level constraint? Validators are now injectable so you can create a custom validator and inject into that. There are lots of resources on how to do this, e.g. http://musingsofaprogrammingaddict.blogspot.com/2010/02/generic-class-level-constraint-for-bean.html (I haven't tried it but in theory it should work).

As long as class-level constraints can do the same thing, I actually think it'd be better to remove the custom behavior of the validate method altogether. It's not clear to me why it's necessary, and I think it's good practice to have some decoupling between the data and the validation.

**Comment by eximius313** [comment]
Ok, moving everything from ```public List<ValidationError> validate()``` to class-level validator could be a solution, but what is the purpose of ```validate()``` then?

**Comment by eximius313** [comment]
And another example - how to return localised message?
With ```validate()``` + DI, one could do:
```
	public List<ValidationError> validate() {
		if(chosenElements.size() > MAX_SIZE){
			return Arrays.asList(new ValidationError(StringUtils.EMPTY, messages.at("limit.exceeded", MAX_SIZE)));
		}
		return null;
	}
```
Is it possible with class-level validator as well?

**Comment by gmethvin** [comment]
You can do the same thing with DI in a validator (inject MessagesApi). Or you could pass the message keys back in the arguments of the ValidationError.

**Comment by eximius313** [comment]
@gmethvin but what you are saying, and even the link you have provided, describes dirty workaround for disadvantages of class-level validator. From the 3rd paragraph:
```
1)You'll probably need a dedicated constraint for every reasonably complex business object. Providing an annotation, a validator implementation and an error message for each can become pretty tedious.

2)Business objects typically know best themselves, whether they are in a consistent, valid state or not. By putting validation logic into a separate constraint validator class, objects have to expose their internal state, which otherwise might not be required.
```
So why to bother with workarounds and not **just** introduce it in Forms (as it is already introduced in Validators)?

**Comment by gmethvin** [comment]
@eximius313 The `@SelfValidating` constraint essentially does the same thing as the `validate` method in Play forms, which eliminates this problem in the specific case mentioned, where the business object knows already if it's in an internal state. So I guess it makes sense for Play to provide one or the other.

You're trying to validate based on external state (your user repository), which as we discussed is not really a typical use of form validation. My point is that it's still possible to do despite the limitations.

**Comment by eximius313** [comment]
Ok, but what is the argument **against** providing DI in Forms (if it is already present in validators, indicating the need of using external state during validation anyway)?

**Comment by gmethvin** [comment]
I'm not sure the implementation would be trivial, and work with every DI framework implementation, since those classes will have some components injected and some not. Most model classes people are using with forms aren't designed to be created with DI.

A lot of the forms code hasn't been touched in a while, and don't think any of the Play core developers that familiar with it. Before we start making tons of changes I would prefer to fully evaluate the options. I would be happier with a more pure JSR-303 approach based on validators than the "custom" solution we have now.

**Comment by marcospereira** [comment]
> Ok, moving everything from public List<ValidationError> validate() to class-level validator could be a solution, but what is the purpose of validate() then?

Maybe deprecated and remove it later? What do you think? Of course, better docs are required here.

>  I would be happier with a more pure JSR-303 approach based on validators than the "custom" solution we have now.

@gmethvin Not sure what you mean by "more pure", but [current approach already uses JSR-303](https://www.playframework.com/documentation/2.5.x/JavaForms#Defining-constraints). Thanks to @mkurz, `play.data.Form` objects now supports DI when created using `play.data.FormFactory`, but keep in mind that this is different than inject dependencies into the object held by the `Form` object. Also, looks like custom validators can use DI already (take a look at `play.data.validation.DefaultConstraintValidatorFactory` which is bounded at `play.inject.BuiltInModule`).

**Comment by gmethvin** [comment]
Calling the validate method is a hack that was added on top in Play Forms. It's not a standard JSR-303 thing as far as I know.

**Comment by eximius313** [comment]
@marcospereira so what I'm asking for is to go one step further and allow DI in ```Form``` Object

**Comment by mkurz** [comment]
I also used the `validate` method when starting with Play 2, but now I try to avoid it if that's possible (once I was checking if two passwords match - so no injected component was needed). I probably would prefer to use something like `@SelfValidating` now (and will for new projects).

One disadvantage of using the `validate` method also is/was that it gets triggered only if all other constraints pass in the first place. E.g if you submit a signup form with "username", "email", "email-confirmation" fields and you fail the checks for the "username" field (because e.g. you exceed the `@MaxLength` annotation) you will *only* see the validation error message for the "username" field - even if you already entered two non-matching email addresses. Only after fixing the failing constraints and re-submitting the form the `validate` method gets triggerd and the user sees the validation errors for the emails fields. Having this second step was/is quite annoying because it could have been displayed in the first place already. Such a user experience is just horrible (User expects he/she fixed the form errors, re-submits -> gets new errors on fields that had no errors before).
I guess using something like `@SelfValidating` would fix this.

Also I don't really see the need of injecting components into the form itself. As said before I implemented DI for validation (and also formatters) - so switching away from `validate` to using class level constraints should do the trick. As it was mentioned earlier the form objects are basically used for transporting data (that's how I use them at least).

However:
I had a quick look how we could add DI support for forms and it shouldn't be too difficult.
The `Form` class would need an `Injector` object, which we could pass into the constructors like we do with `messagesApi`, `formatters`, etc. already (via the `FormFactory`).
Then all we need is to do is change [this line](https://github.com/playframework/playframework/blob/2.5.1/framework/src/play-java/src/main/java/play/data/Form.java#L128) to `injector.instanceOf(backedType)`.
To make this work forms need to extend a specific class/interface that we need to introduce and we need to add a module and bind this class/interface (e..g `bind(AbstractForm.class).toSelf()`.
I think that *should* work (again -> "should").
But that's just the theory...

**Comment by mkurz** [comment]
If I find some time the next weeks I can try to provide a PR. (I am quite busy now with a new project for a client of mine.) But I don't want to promise anything.

**Comment by eximius313** [comment]
@mkurz thank you for very good explanation.
I'll also have a look at ```@SelfValidating```

**Comment by marcospereira** [comment]
> Then all we need is to do is change this line to injector.instanceOf(backedType).

@mkurz should we really do this? Looks like to me that `validate` methods have a lot of problems:

1. They are a hack around JSR-303 as pointed out by @gmethvin
2. They didn't work well with other constraints, like you explained
3. They are blocking, which is not self evident, but accessing external dependencies inside these methods will turn actions that do validation into blocking actions.

My opinion here is that we should deprecated `validate` in favor of using plain JSR-303 (with some custom constrains of our own) and ensure that implementations of `ConstraintValidator` supports DI. This will not solve problem 3 though.

**Comment by mkurz** [comment]
@marcospereira I agree. I also think we should deprecate `validate`.
However even if we remove `validate`, we could still use the injector to create the new form instances - there is nothing wrong to do so, or what do you think?

**Comment by gmethvin** [comment]
@mkurz If we don't have `validate`, is there any benefit to using the injector to create the actual backed class instance? Normally this is just a data class, so it's not a component you'd generally need to be injected. The main place you would use it is in validation.

**Comment by mkurz** [comment]
OK, then we just leave it as it is and deprecate `validate`.

**Comment by gmethvin** [comment]
Another possibility is allowing a user-provided `Supplier<T> newInstance` that would create the "blank" instance of the class. That way Play doesn't need to be opinionated about how that instance is created, and it enables other useful things like using classes that don't have default constructors.

**Comment by 0xbaadf00d** [comment]
Hi,

whatever the choice you make, I think you have to keep in mind two points:

-  They are 2 possible usage flows :

    1. Use form only to bind data and all verifications are done later in controller
    2. Use form to validate everything, and so, in controler, after xxx.hasError() we are definitevely sure submitted data are correct and valid. We can use controller to only do intended logic

- When a field (ie: email) is not valid (due to Contraints.Email or Constraints.MaxLength), the value must be NULL when accessing this field from another validator. Because no one want rewrite xxxx time the same check when we have to validate two field together.

**Comment by jroper** [comment]
My 2 cents, I don't think the validate method should be deprecated.  It's incredibly convenient for doing validations across fields, for example, if a form has a field for entering a minimum value and another field for entering a maximum value, you want to validate that the maximum value is greater than the minimum.  Or, if you have a radio box, with an option for other, you want to validate that the text field associated with the other radio box is non empty but only if other was selected.

Now I'm guessing (I don't know for sure) that you can probably define a custom validator to do this and apply it at the class level, but it will be specific to your form, it won't be reusable.  So why would you want to go through all the boilerplate of implementing the custom annotation when you're only going to use it in one place?  The validate method I think isn't a hack, I think it's very important, it's how many other MVC frameworks do multi field validation.

**Comment by schmitch** [comment]
I guess the big problem of this issue is how data should be modeled.
Even in other Frameworks a lot of the time you see example's where people directly do validation in their form, i.e. adding database values to their forms.
But I guess that's not a good practice since it's not really useful to query the database at all if the validation fails at their basic level. (email is not a email, number is a string, etc).

I guess RoR and Django actually have ModelForms, where you directly attach your database model to your validation data, while this sounds simple at first, it will be a whole mess once your application grows. 
You will end up with everything needs database access, which starts to be a uncontrollable mess pretty fast.

**Comment by mkurz** [comment]
@jroper You are wrong: You actually **can** implement re-usable (general) class level validators. That's what we do in our Project. We only implemented one class-level validator which we apply on several different forms.
We basically end up having the same result like when using the `validate` method: Forms which need cross-field validation (or with similiar concerns) implement an interface which defines an abstract method `isValid` which we override in the forms. That's exactly the same like having the `validate` method - except that you need to implement an interface and annotate the form with the class-level validator.

Please read [my comment above](https://github.com/playframework/playframework/issues/5992#issuecomment-206262170) why the `validate` method isn't working very well (and because of this we don't use it and I don't recommend anyone using it).

My suggestion is to provide a "default" class-level validator in play-java which people can use (again, that's almost the same as having the current `validate` method, not much different) OR/AND to document how to setup up such a validator.

Class level validators are exactly what the `validate` method tries to mimic.
That's why I still think we should deprecate the current `validate` method.

I am quite busy right now, but If I find time I will post our implemention with an example and maybe set up a PR.

**Comment by mkurz** [comment]
Also you can include a class-level validator in form groups which you can't do with the current `validate` method.

**Comment by mkurz** [comment]
@0xbaadf00d
> Actually (Play 2.5.3), we have overridded your Form to get ability to process "partial form" (the form is pre-filled with other data and use bindFromRequest to update only provided fields), re-set to NULL field not validated by constraints and call validate method each time (even if all constraints are not satisfied).

Actually this is a dirty hack - you can use form groups to implement "partial forms". (But sadly, as a said before in a different comment, the `validate` method can't be part of such any form group - but a class level validator's `isValid` method could be).

**Comment by 0xbaadf00d** [comment]
Class level validators are complicated to use, and your document not contains enough example**s** . The method `validate` is realy usefull to do complex validation (eg: we override your built-in class to allow templated argument to, by example, avoid money transfer from one account to the same account who emit the request).

In fact, this method must stay as in, if you dont want use it: dont use it, but let's users who want use it, using it.

**Comment by mkurz** [comment]
Class level validators are not complicated to use. As I said I will post an implementation (unfortunatly I have a deadline to meet until tomorrow morning before a phone conference).
In fact, if we provide a default validator in play-java (which basically is an interface and an annotation with it's constraint) all you would have to do for migration is to implement the interface and annote the form with let says `@SelfValidating` (or however we call it). You could then just re-use your existing `validate` method. We also could name the method of the interface `validate` - so you end up almost the same like you have now. Not very complicated.
Actually I don't see any problem at all - you don't need to fear you loose any simplicity. It's stays as simple as it is now. But you gain a lot of flexibility (support for form groups).

Which class to you override? Form?
To be honest I think this isn't a very good idea. I strongly recommend you to use form groups which are fully working in Play 2.5.8.

Having heavily worked on Play's Form implemention (fixing quite some bugs, providing DI for validators, formatters, etc.) and given that I know when and how the validate method is called I am very confident that we should remove the `validate` method and instead use/provide class-level validators.

**Comment by 0xbaadf00d** [comment]
I wait and see.  Currently we dont bind directly models (and we don't, never) but some time, validator method need some model objects instance (ie: current connected user, current contract, ...) If your "Class Level Validator" allow this...  why not.  But When our developers  pass the `if(!myForm.hasError()) {}` all data has been checked. No check / validation on controller directly.

**Comment by mkurz** [comment]
This will be fixed by #7038 which provides a pure `JSR-303`/`JSR-349` approach of validating the state of an entire object via class-level constraints.

All concerns mentioned in this issue are addressed in my pull request:
* Cross field validation
* Partial form validation
* For most use cases you don't have to write your own class-level constraints because Play provides a basic set of generic ones to easily replace the current `validate` method
* You can inject components like a database into class-level constraints
* The order of validation (e.g. basic field validation (required, min, max, etc.) should happen before validation which e.g. hits the database)
* After calling `myForm.hasErrors()` in a controller everything is validated - you don't have to do any validation in controllers anymore (or hacks like sub-class Play's `Form` class...)

Migrating an existing `validate` method to Play's built-in class level constraints is easy.
Depending on the return type of the `validate` method a form class like:
```java
public class MyForm {
    //...
    public String validate() {
        //...
    }
}
```
will become:
```java
import play.data.validation.Constraints.SelfValidatingSimple;
import play.data.validation.Constraints.ValidatableSimple;

@SelfValidatingSimple
public class MyForm implements ValidatableSimple {
    //...
    @Override
    public String validateInstance() {
        //...
    }
}
```

Have a look at the docs and examples I added in #7038 which should explain all use cases and how to write you own class-level constraints if needed.
